pegen/metagrammar.gram

```diff
@@ -0,0 +1,55 @@
+start: rules ENDMARKER {Rules(rules)}
+
+rules:
+    | rule rules { [rule] + rules }
+    | rule { [rule] }
+
+rule:
+    | rulename ":" alts NEWLINE INDENT more_alts DEDENT { Rule(*rulename, alts + more_alts) }
+    | rulename ":" NEWLINE INDENT more_alts DEDENT { Rule(*rulename, more_alts) }
+    | rulename ":" alts NEWLINE { Rule(*rulename, alts) }
+
+rulename:
+        | NAME '[' type=NAME '*' ']' {(name.string, type.string+"*")}
+        | NAME '[' type=NAME ']' {(name.string, type.string)}
+        | NAME {(name.string, None)}
+
+alts:
+    | alt "|" alts { Rhs([alt] + alts.alts)}
+    | alt { Rhs([alt]) }
+
+more_alts:
+    | "|" alts NEWLINE more_alts { Rhs(alts.alts + more_alts.alts) }
+    | "|" alts NEWLINE { Rhs(alts.alts) }
+
+alt:
+    | items ['$'] action { Alt(items + ([NamedItem(None, NameLeaf('ENDMARKER'))] if opt else []), action=action.string) }
```

Comment:
I would give an explicit name to the `['$']` piece.

Also the action is hard to follow -- maybe just write out two alternatives, `item '$' action` and `item action`? (And two more for the variation without `action`.)